clang/lib/Sema/SemaDecl.cpp

```diff
@@ -16232,7 +16232,9 @@ Decl *Sema::ActOnFinishFunctionBody(Decl *dcl, Stmt *Body,
 
       // If the function implicitly returns zero (like 'main') or is naked,
       // don't complain about missing return statements.
-      if (FD->hasImplicitReturnZero() || FD->hasAttr<NakedAttr>())
+      if ((FD->hasImplicitReturnZero() &&
+           (getLangOpts().CPlusPlus || getLangOpts().C99 || !FD->isMain())) ||
```

Comment:
This change looks wrong to me. We shouldn't have to look at the language options and whether the function is or isn't `main`, the point is to rely on `hasImplicitReturnZero()` alone. Pretty sure the fallthrough checker is what should be handling silencing the diagnostic.

Reply:
@AaronBallman, thanks for the feedback. Do you mean we should avoid setting setHasImplicitReturnZero for C89?

https://github.com/llvm/llvm-project/blob/0ab2061c4fd0feac8b142ff76ed823534bebc634/clang/lib/Sema/SemaDecl.cpp#L12445

Reply:
No, I think we should keep the behavior we have, it's that I think the analysis pass should be run and the diagnostic be suppressed from within there.

Reply:
Oh I see what's going on, thank you for the explanation. Then I think all that's needed here is a comment saying something along the lines of: Clang implicitly returns 0 in C89 mode, but that's an extension, so we need to perform the check in that case to ensure we emit the expected extension warning in C89 mode.

Reply:
@AaronBallman I've added comment for these changes